File: training/easy/defibrillators.kt
import java.lang.Math.PI
import java.lang.Math.cos
import java.util.*

fun Double.toRad() = this * PI / 180
fun String.toFixedDouble() = this.replace(',', '.').toDouble()

data class Coord(val lon: Double, val lat: Double) {
    fun distance(other: Coord): Double {
        val x = (lon - other.lon) * cos((lat.toRad() + other.lat.toRad()) / 2)
        val y = lat - other.lat
        return x * x + y * y
    }
}


fun main(args : Array<String>) {
    val input = Scanner(System.`in`)
    val user = Coord(input.next().toFixedDouble(), input.next().toFixedDouble())
    val n = input.nextInt()
    input.nextLine()

    val (res, _) = (0 until n).fold("" to Double.MAX_VALUE) {
            (name, d), _ ->
        val def = input.nextLine().split(';')
        val dist = user.distance(Coord(def[4].toFixedDouble(), def[5].toFixedDouble()))
        if (dist < d) def[1] to dist else name to d
    }

    println(res)
}

File: training/easy/defibrillators.scala
import scala.io.StdIn._

object Defiblirators extends App {
  case class Loc(lon: Double, lat: Double)
  case class Defib(name: String, loc: Loc)
  case class Result(name: String, dist: Double)

  val lon = readLine
  val lat = readLine
  val loc = Loc(lon.replace(',','.').toDouble, lat.replace(',','.').toDouble)
  println((0 until readInt).foldLeft(Result("", Double.MaxValue))(process).name)

  def process(res: Result, _n: Int): Result = {
    val defib = getNameAndLoc(readLine)
    val dist = calculateDistance(loc, defib.loc)
    if (dist < res.dist) Result(defib.name, dist) else res
  }

  def getNameAndLoc(defib:String): Defib = {
    val data = defib.split(";")
    val lon = data(4).replace(',','.').toDouble
    val lat = data(5).replace(',','.').toDouble
    Defib(data(1), Loc(lon, lat))
  }

  def calculateDistance(a: Loc, b: Loc): Double = {
    val x = (b.lon - a.lon) * math.cos ((a.lat + b.lat) / 2)
    val y = b.lat - a.lat
    x * x + y * y
  }
}
